text/0000-grid-primitives.md

```diff
@@ -0,0 +1,761 @@
+    data_type=PropertyDefinition.DataType.NUMBER,
+    number_exponent=-2,
+    required=True,
+    description="The the price this object"
+)
+
+```
+
+This example shows an instance of a price with the value $0.23:
+
+```
+PropertyValue(
+    name="price",
+    data_type=PropertyDefinition.DataType.NUMBER,
+    number_value=23,
+)
+```
+
+Because this is a protobuf message, the default exponent is `0` when the schema
+is created. Likewise, the default value for this property instance is `0`.
+
+### Enum
+
+An enum value would be represented as:
+
+```
+PropertyDefinition(
+    name='color',
+    data_type=PropertyDefinition.DataType.ENUM,
+    enum_options=['white', 'red', 'green', 'blue', 'blacklight'],
+    required=True
+)
+``` 
+
+An instance of this enum, indicating "red", would be as follows:
+
+```
+PropertyValue(
+    name='color',
+    data_type=PropertyDefinition.DataType.ENUM,
+    enum_value=1
+)
+```
+
+Due to the use of protobuf, the default value for this property instance is `0`.
+It is left to the smart-contract implementer to determine if this would result
+in an error condition.
+
+### Struct
+
+A struct value would be represented as follows:
+
+```
+PropertyDefinition(
+    name='shock',
+    data_type=PropertyDefinition.DataType.STRUCT,
+    struct_properties=[
+        PropertyDefinition(
+            name='speed',
+            data_type=PropertyDefinition.DataType.NUMBER,
+            number_exponent=-6),
+        PropertyDefinition(
+            name='duration',
+            data_type=PropertyDefinition.DataType.NUMBER,
+            number_exponent=-6),
+    ],
+    required=True
+)
+```
+
+An instance of the “shock" struct would be as follows:
+
+```
+PropertyValue(
+    name='shock',
+    data_type=PropertyDefinition.DataType.STRUCT,
+    struct_values=[
+        PropertyValue(
+            name='speed',
+            data_type=PropertySchema.DataType.NUMBER,
+            number_value=500000),
+        PropertyValue(
+            name='duration',
+            data_type=PropertySchema.DataType.NUMBER,
+            number_value=10000)
+        ])
+```
+
+The property value for a struct must contain all the struct values from the
+property definition, or it is invalid.  The defaults for the struct values
+themselves depend on their data types and/or the smart-contract implementer
+validation rules.
+
+### Latitude/Longitude
+
+A latitude/longitude (lat/long) value would be represented as follows:
+
+```
+PropertyDefintion(
+    name='origin',
+    data_type=PropertyDefinition.DataType.LAT_LONG,
+    required=True
+)
+```
+
+A lat/long instance would be as follows:
+
+```
+PropertyValue(
+    name='origin',
+    data_type=PropertyDefinition.DataType.LAT_LONG,
+    lat_long_value=LatLong(
+        latitude=44977753,
+        longitude=-93265015)
+)
+```
+
+Due to the use of protobuf, the default values for `LatLong` would be `(0, 0)`.
+While this is a valid lat/long, it could be used to indicate an error, depending
```

Comment:
Would this ever be used to indicate an error? Seems bad for a smart contract to assume that a valid value is an error (unless intentionally restricting the space of valid coordinates).

Reply:
0, 0 could be considered a unrealistic measurement, given that it is highly unlikely that a device would ever read that as a value.  That being said, the RFC leaves it up to the contract developer to make that call - it has no opinions here.

Reply:
(0,0) is in the Gulf of Guinea; possible that a ship passing through this location could send this as a readout of (0,0) I am with @Itseeley on this one-

Reply:
> Due to the use of protobuf, the default values for LatLong would be (0, 0). While this is a valid lat/long, it could be used to indicate an error, **depending on the choice of the smart-contract implementer**.

Again, this piece of validation is left to the smart-contract developer.  It's not a schema operation.

Keep in mind, the lat long is in millionths of degrees, which is on the order of single-digit feet.